File: wwwroot/src/raytracer/vec3.ts
import { rand } from "./numberUtils";

export interface Vec3 {
    x: number;
    y: number;
    z: number;
}

export function add(v1: Vec3, v2: Vec3): Vec3 {
    return {
        x: v1.x + v2.x,
        y: v1.y + v2.y,
        z: v1.z + v2.z
    };
}

export function subtract(v1: Vec3, v2: Vec3): Vec3 {
    return {
        x: v1.x - v2.x,
        y: v1.y - v2.y,
        z: v1.z - v2.z
    };
}

export function multiply(v1: Vec3, v2: Vec3): Vec3 {
    return {
        x: v1.x * v2.x,
        y: v1.y * v2.y,
        z: v1.z * v2.z,
    }
}

export function scale(v: Vec3, scale: number): Vec3 {
    return {
        x: v.x * scale,
        y: v.y * scale,
        z: v.z * scale
    }
}

export function dot(v1: Vec3): number
export function dot(v1: Vec3, v2: Vec3): number
export function dot(v1: Vec3, v2?: Vec3): number {
    if (!v2) v2 = v1;

    return v1.x * v2.x +
        v1.y * v2.y +
        v1.z * v2.z;
}

export function cross(v1: Vec3, v2: Vec3): Vec3 {
    return {
        x: v1.y * v2.z - v1.z * v2.y,
        y: v1.z * v2.x - v1.x * v2.z,
        z: v1.x * v2.y - v1.y * v2.x
    };
}

export function reflect(v: Vec3, normal: Vec3): Vec3 {
    return subtract(v, scale(normal, 2 * dot(v, normal)));
}

export function refract(v: Vec3, normal: Vec3, refractionIndexRatio: number) {
    const cosTheta = Math.min(dot(scale(v, -1), normal), 1.0);
    const perpendicular = scale(add(v, scale(normal, cosTheta)), refractionIndexRatio);
    const parallel = scale(normal, -Math.sqrt(Math.abs(1.0 - dot(perpendicular))));
    return add(perpendicular, parallel);
}

export function length(v: Vec3): number {
    return Math.sqrt(dot(v));
}

export function normalise(v: Vec3): Vec3 {
    return scale(v, 1 / length(v));
}

export function nearZero(v: Vec3): boolean {
    const epsilon = 1e-8;
    return Math.abs(v.x) < epsilon
        && Math.abs(v.y) < epsilon
        && Math.abs(v.z) < epsilon;
}

export function lerp(v1: Vec3, v2: Vec3, t: number): Vec3 {
    return add(scale(v1, (1.0 - t)), scale(v2, t))
}

export function toString(v: Vec3) {
    return `(${v.x}, ${v.y}, ${v.z})`;
}

export function randomVec3(min?: number, max?: number) {
    return {
        x: rand(min, max),
        y: rand(min, max),
        z: rand(min, max)
    };
}

export function randomVec3InUnitSphere() {
    while (true) {
        const v = randomVec3(-1, 1);
        if (dot(v) < 1) return v;
    }
}

export function randomVec3InHemisphere(normal: Vec3) {
    const inUnitSphere = randomVec3InUnitSphere();
    if (dot(inUnitSphere, normal) >= 0.0) return inUnitSphere;
    else return scale(inUnitSphere, -1);
}

File: rust/src/vec3.rs
use std::{
    fmt::{Display, Formatter},
    ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign},
};

use crate::utils;

#[derive(Clone, Copy)]
pub struct Vec3 {
    pub x: f64,
    pub y: f64,
    pub z: f64,
}

pub type Point3 = Vec3;

impl Vec3 {
    pub fn origin() -> Vec3 {
        Vec3 {
            x: 0.0,
            y: 0.0,
            z: 0.0,
        }
    }

    pub fn random() -> Vec3 {
        Vec3 {
            x: utils::random(),
            y: utils::random(),
            z: utils::random(),
        }
    }

    pub fn range_random(min: f64, max: f64) -> Vec3 {
        Vec3 {
            x: utils::range_random(min, max),
            y: utils::range_random(min, max),
            z: utils::range_random(min, max),
        }
    }

    pub fn random_unit_vector() -> Vec3 {
        loop {
            let v = Vec3::random();
            let length_squared = v.length_squared();

            if length_squared == 1.0 {
                return v;
            }
            else if length_squared > f64::MIN_POSITIVE && length_squared < 1.0 {
                return v / length_squared.sqrt();
            }
        }
    }

    pub fn random_on_hemisphere(normal: Vec3) -> Vec3 {
        let on_unit_sphere = Vec3::random_unit_vector();

        if on_unit_sphere.dot(normal) > 0.0 {
            // In the same hemisphere as normal
            on_unit_sphere
        } else {
            -on_unit_sphere
        }
    }

    pub fn random_in_unit_disk() -> Vec3 {
        loop {
            let p = Vec3 {
                x: utils::range_random(-1.0, 1.0),
                y: utils::range_random(-1.0, 1.0),
                z: 0.0
            };

            if p.length_squared() < 1.0 {
                return p;
            }
        }
    }

    pub fn length(self) -> f64 {
        self.length_squared().sqrt()
    }

    pub fn length_squared(self) -> f64 {
        self.x * self.x + self.y * self.y + self.z * self.z
    }

    pub fn normalize(self) -> Vec3 {
        self / self.length()
    }

    pub fn near_zero(self) -> bool {
        const S: f64 = 1e-8;

        self.x.abs() < S && self.y.abs() < S && self.z.abs() < S
    }

    pub fn reflect(self, normal: Vec3) -> Vec3 {
        self - 2.0 * self.dot(normal) * normal
    }

    pub fn refract(self, normal: Vec3, refraction_index: f64) -> Vec3 {
        let cos_theta = -self.dot(normal).min(1.0);
        let r_out_perp = refraction_index * (self + cos_theta * normal);
        let r_out_para = -((1.0 - r_out_perp.length_squared()).abs().sqrt()) * normal;

        r_out_perp + r_out_para
    }

    pub fn dot(self, rhs: Vec3) -> f64 {
        self.x * rhs.x + self.y * rhs.y + self.z * rhs.z
    }

    pub fn cross(self, rhs: Vec3) -> Vec3 {
        Vec3 {
            x: self.y * rhs.z - self.z * rhs.y,
            y: self.z * rhs.x - self.x * rhs.z,
            z: self.x * rhs.y - self.y * rhs.x,
        }
    }
}

impl Display for Vec3 {
    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
        write!(f, "({}, {}, {})", self.x, self.y, self.z)
    }
}

impl Add for Vec3 {
    type Output = Vec3;

    fn add(self, rhs: Vec3) -> Self::Output {
        Vec3 {
            x: self.x + rhs.x,
            y: self.y + rhs.y,
            z: self.z + rhs.z,
        }
    }
}

impl AddAssign for Vec3 {
    fn add_assign(&mut self, rhs: Vec3) {
        self.x += rhs.x;
        self.y += rhs.y;
        self.z += rhs.z;
    }
}

impl Div<f64> for Vec3 {
    type Output = Vec3;

    fn div(self, rhs: f64) -> Self::Output {
        Vec3 {
            x: self.x / rhs,
            y: self.y / rhs,
            z: self.z / rhs,
        }
    }
}

impl DivAssign<f64> for Vec3 {
    fn div_assign(&mut self, rhs: f64) {
        self.x /= rhs;
        self.y /= rhs;
        self.z /= rhs;
    }
}

impl Mul for Vec3 {
    type Output = Vec3;

    fn mul(self, rhs: Vec3) -> Self::Output {
        Vec3 {
            x: self.x * rhs.x,
            y: self.y * rhs.y,
            z: self.z * rhs.z,
        }
    }
}

impl Mul<f64> for Vec3 {
    type Output = Vec3;

    fn mul(self, rhs: f64) -> Self::Output {
        Vec3 {
            x: self.x * rhs,
            y: self.y * rhs,
            z: self.z * rhs,
        }
    }
}

impl Mul<Vec3> for f64 {
    type Output = Vec3;

    fn mul(self, rhs: Vec3) -> Vec3 {
        rhs * self
    }
}

impl MulAssign for Vec3 {
    fn mul_assign(&mut self, rhs: Vec3) {
        self.x *= rhs.x;
        self.y *= rhs.y;
        self.z *= rhs.z;
    }
}

impl Neg for Vec3 {
    type Output = Vec3;

    fn neg(self) -> Self::Output {
        Vec3 {
            x: -self.x,
            y: -self.y,
            z: -self.z,
        }
    }
}

impl Sub for Vec3 {
    type Output = Vec3;

    fn sub(self, rhs: Vec3) -> Self::Output {
        Vec3 {
            x: self.x - rhs.x,
            y: self.y - rhs.y,
            z: self.z - rhs.z,
        }
    }
}

impl SubAssign for Vec3 {
    fn sub_assign(&mut self, rhs: Vec3) {
        self.x -= rhs.x;
        self.y -= rhs.y;
        self.z -= rhs.z;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn origin_is_000()
    {
        let origin = Vec3::origin();

        assert_eq!(0.0, origin.x);
        assert_eq!(0.0, origin.y);
        assert_eq!(0.0, origin.z);
    }

    #[test]
    fn should_fail()
    {
        assert_eq!(5, 2+2)
    }
}
